kernel/aster-nix/src/process/sync/condvar.rs

```diff
@@ -0,0 +1,173 @@
+        let lock = guard_lock(&guard);
+        drop(guard);
+        let cond = || {
+            let wait_list = self.should_wait.lock();
+            if let Some(wait) = wait_list.front() {
+                if wait.load(Ordering::SeqCst) {
+                    return None;
+                }
+            }
+
+            Some(())
+        };
+        // Wait until the condition becomes true, we're explicitly woken up, or the timeout elapses.
+        let res = self.pauser.pause_until_or_timeout(cond, &timeout);
+        self.should_wait.lock().pop_front();
+        match res {
+            Ok(_) => Ok((lock.lock(), false)),
+            Err(error) => match error.error() {
+                Errno::EINTR => Err(LockErr::Interrupted((lock.lock(), false))),
+                Errno::ETIME => Err(LockErr::Timeout((lock.lock(), true))),
+                _ => Err(LockErr::Unknown((lock.lock(), false))),
+            },
+        }
+    }
+
+    /// Wait for the condition to become true, or until the timeout elapses, or until the condition is explicitly woken up.
+    /// bool is true if the timeout is reached.
+    pub fn wait_timeout_while<'a, T, F>(
+        &self,
+        mut guard: MutexGuard<'a, T>,
+        timeout: Duration,
+        mut condition: F,
+    ) -> LockResult<(MutexGuard<'a, T>, bool)>
+    where
+        F: FnMut(&mut T) -> bool,
+    {
+        loop {
+            if !condition(&mut *guard) {
+                return Ok((guard, false));
+            }
+            guard = match self.wait_timeout(guard, timeout) {
+                Ok((guard, timeout_flag)) => guard,
+                Err(LockErr::Timeout((guard, timeout_flag))) => {
+                    return Err(LockErr::Timeout((guard, timeout_flag)))
+                }
+                Err(LockErr::Interrupted(guard)) => return Err(LockErr::Interrupted(guard)),
+                Err(LockErr::Unknown(guard)) => return Err(LockErr::Unknown(guard)),
+            }
+        }
+    }
+
+    // Wait for the condition to become true, and until the condition is explicitly woken up or interupted.
+    pub fn wait_while<'a, T, F>(
+        &self,
+        mut guard: MutexGuard<'a, T>,
+        mut condition: F,
+    ) -> LockResult<MutexGuard<'a, T>>
+    where
+        F: FnMut(&mut T) -> bool,
+    {
+        loop {
+            if !condition(&mut *guard) {
+                return Ok(guard);
+            }
+            guard = match self.wait(guard) {
+                Ok(guard) => guard,
+                Err(LockErr::Interrupted(guard)) => return Err(LockErr::Interrupted(guard)),
+                Err(LockErr::Unknown(guard)) => return Err(LockErr::Unknown(guard)),
+                _ => unreachable!(),
+            }
+        }
+    }
+
+    /// Wakes up one task waiting on the condition variable.
+    pub fn notify_one(&self) {
+        if let Some(x) = self.should_wait.lock().front() {
+            x.store(false, Ordering::SeqCst)
+        }
+        self.pauser.resume_one();
```

Comment:
Well, this method requires Pauser also resumes the front waiter, but indeed the pauser does not provides this property. But I cannot think up a better solution, so just add a FIXME comment here

Reply:
All right. So `WaitQueue` cannot also wake up all waiters?